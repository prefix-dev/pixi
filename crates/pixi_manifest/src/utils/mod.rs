mod spanned;

pub use spanned::PixiSpanned;
use url::Url;

/// If the URL points to a subdirectory, extract it, as in (git):
///   `git+https://git.example.com/MyProject.git@v1.0#subdirectory=pkg_dir`
///   `git+https://git.example.com/MyProject.git@v1.0#egg=pkg&subdirectory=pkg_dir`
/// or (direct archive url):
///   `https://github.com/foo-labs/foo/archive/master.zip#subdirectory=packages/bar`
///   `https://github.com/foo-labs/foo/archive/master.zip#egg=pkg&subdirectory=packages/bar`
pub(crate) fn extract_directory_from_url(url: &Url) -> Option<String> {
    let fragment = url.fragment()?;
    let subdirectory = fragment
        .split('&')
        .find_map(|fragment| fragment.strip_prefix("subdirectory="))?;
    Some(subdirectory.into())
}

#[cfg(test)]
mod test {
    use rstest::*;
    use url::Url;

    use super::*;

    #[rstest]
    #[case(
        "git+https://github.com/foobar.git@v1.0#subdirectory=pkg_dir",
        Some("pkg_dir")
    )]
    #[case(
        "git+ssh://gitlab.org/foobar.git@v1.0#egg=pkg&subdirectory=pkg_dir",
        Some("pkg_dir")
    )]
    #[case("git+https://github.com/foobar.git@v1.0", None)]
    #[case("git+https://github.com/foobar.git@v1.0#egg=pkg", None)]
    #[case(
        "git+https://github.com/foobar.git@v1.0#subdirectory=pkg_dir&other=val",
        Some("pkg_dir")
    )]
    #[case(
        "git+https://github.com/foobar.git@v1.0#other=val&subdirectory=pkg_dir",
        Some("pkg_dir")
    )]
    #[case(
        "git+https://github.com/foobar.git@v1.0#subdirectory=pkg_dir&subdirectory=another_dir",
        Some("pkg_dir")
    )]
    fn test_get_subdirectory(#[case] url: Url, #[case] expected: Option<&str>) {
        let subdirectory = extract_directory_from_url(&url);
        assert_eq!(subdirectory.as_deref(), expected);
    }
}
