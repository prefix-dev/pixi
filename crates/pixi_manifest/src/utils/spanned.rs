//! Taken from: <https://docs.rs/serde_spanned/latest/serde_spanned/struct.Spanned.html>
use std::cmp::Ordering;
use std::hash::{Hash, Hasher};

// Currently serde itself doesn't have a spanned type, so we map our `Spanned`
// to a special value in the serde data model. Namely one with these special
// fields/struct names.
//
// In general, supported deserializers should catch this and not literally emit
// these strings but rather emit `Spanned` as they're intended.
#[doc(hidden)]
pub const NAME: &str = "$__serde_spanned_private_Spanned";
#[doc(hidden)]
pub const START_FIELD: &str = "$__serde_spanned_private_start";
#[doc(hidden)]
pub const END_FIELD: &str = "$__serde_spanned_private_end";
#[doc(hidden)]
pub const VALUE_FIELD: &str = "$__serde_spanned_private_value";

/// A spanned value, indicating the range at which it is defined in the source.
#[derive(Clone, Debug)]
pub struct PixiSpanned<T> {
    /// Byte range
    pub span: Option<std::ops::Range<usize>>,
    /// The spanned value.
    pub value: T,
}

impl<T: Default> Default for PixiSpanned<T> {
    fn default() -> Self {
        Self {
            span: None,
            value: T::default(),
        }
    }
}

impl<T> From<T> for PixiSpanned<T> {
    fn from(value: T) -> Self {
        Self { span: None, value }
    }
}

impl<T> PixiSpanned<T> {
    /// Byte range
    pub fn span(&self) -> Option<std::ops::Range<usize>> {
        self.span.clone()
    }

    /// Consumes the spanned value and returns the contained value.
    pub fn into_inner(self) -> T {
        self.value
    }

    /// Returns a reference to the contained value.
    pub fn get_ref(&self) -> &T {
        &self.value
    }

    /// Returns a mutable reference to the contained value.
    pub fn get_mut(&mut self) -> &mut T {
        &mut self.value
    }
}

impl std::borrow::Borrow<str> for PixiSpanned<String> {
    fn borrow(&self) -> &str {
        self.get_ref()
    }
}

impl<T> AsRef<T> for PixiSpanned<T> {
    fn as_ref(&self) -> &T {
        self.get_ref()
    }
}

impl<T> AsMut<T> for PixiSpanned<T> {
    fn as_mut(&mut self) -> &mut T {
        self.get_mut()
    }
}

impl<T: PartialEq> PartialEq for PixiSpanned<T> {
    fn eq(&self, other: &Self) -> bool {
        self.value.eq(&other.value)
    }
}

impl<T: Eq> Eq for PixiSpanned<T> {}

impl<T: Hash> Hash for PixiSpanned<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}

impl<T: PartialOrd> PartialOrd for PixiSpanned<T> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.value.partial_cmp(&other.value)
    }
}

impl<T: Ord> Ord for PixiSpanned<T> {
    fn cmp(&self, other: &Self) -> Ordering {
        self.value.cmp(&other.value)
    }
}

impl<'de, T> serde::de::Deserialize<'de> for PixiSpanned<T>
where
    T: serde::de::Deserialize<'de>,
{
    fn deserialize<D>(deserializer: D) -> Result<PixiSpanned<T>, D::Error>
    where
        D: serde::de::Deserializer<'de>,
    {
        struct SpannedVisitor<T>(::std::marker::PhantomData<T>);

        impl<'de, T> serde::de::Visitor<'de> for SpannedVisitor<T>
        where
            T: serde::de::Deserialize<'de>,
        {
            type Value = PixiSpanned<T>;

            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                formatter.write_str("a spanned value")
            }

            fn visit_map<V>(self, mut visitor: V) -> Result<PixiSpanned<T>, V::Error>
            where
                V: serde::de::MapAccess<'de>,
            {
                if visitor.next_key()? != Some(START_FIELD) {
                    return Err(serde::de::Error::custom("spanned start key not found"));
                }
                let start: usize = visitor.next_value()?;

                if visitor.next_key()? != Some(END_FIELD) {
                    return Err(serde::de::Error::custom("spanned end key not found"));
                }
                let end: usize = visitor.next_value()?;

                if visitor.next_key()? != Some(VALUE_FIELD) {
                    return Err(serde::de::Error::custom("spanned value key not found"));
                }
                let value: T = visitor.next_value()?;

                Ok(PixiSpanned {
                    span: Some(start..end),
                    value,
                })
            }
        }

        let visitor = SpannedVisitor(::std::marker::PhantomData);

        static FIELDS: [&str; 3] = [START_FIELD, END_FIELD, VALUE_FIELD];
        deserializer.deserialize_struct(NAME, &FIELDS, visitor)
    }
}

impl<T: serde::ser::Serialize> serde::ser::Serialize for PixiSpanned<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::ser::Serializer,
    {
        self.value.serialize(serializer)
    }
}

#[cfg(test)]
mod tests {
    use crate::utils::spanned::PixiSpanned;
    use serde::Deserialize;

    #[test]
    pub fn test_spanned() {
        #[derive(Deserialize)]
        struct Value {
            s: PixiSpanned<String>,
        }

        let t = "s = \"value\"\n";

        let u: Value = toml_edit::de::from_str(t).unwrap();

        assert_eq!(u.s.span().unwrap(), 4..11);
        assert_eq!(u.s.get_ref(), "value");
        assert_eq!(u.s.into_inner(), String::from("value"));
    }
}
