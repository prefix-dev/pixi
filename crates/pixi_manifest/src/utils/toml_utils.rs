/// Escapes a TOML key if it contains characters that require escaping.
/// According to TOML spec, keys with dots, spaces, or other special characters
/// should be quoted.
pub fn escape_toml_key(key: &str) -> String {
    // Characters that require escaping in TOML keys
    if key.contains([
        '.', ' ', '\t', '\n', '#', '=', '\'', '"', '[', ']', '{', '}',
    ]) {
        format!("\"{}\"", key.replace('\"', "\\\""))
    } else {
        key.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_escape_toml_key() {
        // Keys without special characters don't get modified
        assert_eq!(escape_toml_key("simple"), "simple");
        assert_eq!(escape_toml_key("simple_key"), "simple_key");
        assert_eq!(escape_toml_key("simpleKey"), "simpleKey");

        // Keys with special characters get properly quoted
        assert_eq!(escape_toml_key("test.test"), "\"test.test\"");
        assert_eq!(escape_toml_key("key with space"), "\"key with space\"");
        assert_eq!(escape_toml_key("key.with.dots"), "\"key.with.dots\"");

        // Keys with quotes get properly escaped
        assert_eq!(
            escape_toml_key("key\"with\"quotes"),
            "\"key\\\"with\\\"quotes\""
        );

        // Keys with other special characters
        assert_eq!(escape_toml_key("key#with#hash"), "\"key#with#hash\"");
        assert_eq!(escape_toml_key("key=with=equals"), "\"key=with=equals\"");
        assert_eq!(
            escape_toml_key("key[with]brackets"),
            "\"key[with]brackets\""
        );
    }
}
