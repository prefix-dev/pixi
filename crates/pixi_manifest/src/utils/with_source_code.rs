use std::{
    error::Error,
    fmt::{Debug, Display, Formatter},
};

use miette::{Diagnostic, LabeledSpan, Severity, SourceCode};

/// A struct that binds a [`Diagnostic`] to a [`SourceCode`]. This is similar to
/// using [`miette::Report::with_source_code`] but it retains type information.
#[derive(Debug)]
pub struct WithSourceCode<E, S> {
    pub error: E,
    pub source: S,
}

impl<E: Error + Debug, S: SourceCode + Debug> Error for WithSourceCode<E, S> {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        self.error.source()
    }
}

impl<E: Display, S> Display for WithSourceCode<E, S> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.error)
    }
}

impl<E: Diagnostic, S: SourceCode + Debug> Diagnostic for WithSourceCode<E, S> {
    fn code<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.error.code()
    }

    fn severity(&self) -> Option<Severity> {
        self.error.severity()
    }

    fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.error.help()
    }

    fn url<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.error.url()
    }

    fn source_code(&self) -> Option<&dyn SourceCode> {
        self.error.source_code().or(Some(&self.source))
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
        self.error.labels()
    }

    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn Diagnostic> + 'a>> {
        self.error.related()
    }

    fn diagnostic_source(&self) -> Option<&dyn Diagnostic> {
        self.error.diagnostic_source()
    }
}
