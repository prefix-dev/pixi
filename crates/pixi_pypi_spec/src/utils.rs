use url::Url;

/// Parses URL fragment parameters into key-value pairs.
///
/// URL fragments like `#key1=value1&key2=value2` are parsed into a vector
/// of (key, value) tuples.
///
/// # Examples
/// - `#subdirectory=pkg_dir` -> `[("subdirectory", "pkg_dir")]`
/// - `#sha256=abc123&egg=foo` -> `[("sha256", "abc123"), ("egg", "foo")]`
pub fn parse_url_fragment_parameters(fragment: &str) -> Vec<(&str, &str)> {
    fragment
        .split('&')
        .filter_map(|param| param.split_once('='))
        .collect()
}

/// If the URL points to a subdirectory, extract it, as in (git):
///   `git+https://git.example.com/MyProject.git@v1.0#subdirectory=pkg_dir`
///   `git+https://git.example.com/MyProject.git@v1.0#egg=pkg&subdirectory=pkg_dir`
/// or (direct archive url):
///   `https://github.com/foo-labs/foo/archive/master.zip#subdirectory=packages/bar`
///   `https://github.com/foo-labs/foo/archive/master.zip#egg=pkg&subdirectory=packages/bar`
pub(crate) fn extract_directory_from_url(url: &Url) -> Option<String> {
    let fragment = url.fragment()?;
    parse_url_fragment_parameters(fragment)
        .into_iter()
        .find(|(key, _)| *key == "subdirectory")
        .map(|(_, value)| value.to_string())
}

#[cfg(test)]
mod test {
    use rstest::*;
    use url::Url;

    use super::*;

    #[rstest]
    #[case(
        "git+https://github.com/foobar.git@v1.0#subdirectory=pkg_dir",
        Some("pkg_dir")
    )]
    #[case(
        "git+ssh://gitlab.org/foobar.git@v1.0#egg=pkg&subdirectory=pkg_dir",
        Some("pkg_dir")
    )]
    #[case("git+https://github.com/foobar.git@v1.0", None)]
    #[case("git+https://github.com/foobar.git@v1.0#egg=pkg", None)]
    #[case(
        "git+https://github.com/foobar.git@v1.0#subdirectory=pkg_dir&other=val",
        Some("pkg_dir")
    )]
    #[case(
        "git+https://github.com/foobar.git@v1.0#other=val&subdirectory=pkg_dir",
        Some("pkg_dir")
    )]
    #[case(
        "git+https://github.com/foobar.git@v1.0#subdirectory=pkg_dir&subdirectory=another_dir",
        Some("pkg_dir")
    )]
    fn test_get_subdirectory(#[case] url: Url, #[case] expected: Option<&str>) {
        let subdirectory = extract_directory_from_url(&url);
        assert_eq!(subdirectory.as_deref(), expected);
    }
}
