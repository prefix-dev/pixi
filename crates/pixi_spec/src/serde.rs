use std::{fmt::Display, path::PathBuf};

use rattler_conda_types::{
    BuildNumberSpec, ChannelConfig, NamedChannelOrUrl, NamelessMatchSpec,
    ParseStrictness::{Lenient, Strict},
    StringMatcher, VersionSpec,
};
use rattler_digest::{Md5Hash, Sha256Hash};
use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
use serde_with::serde_as;
use url::Url;

use crate::{DetailedSpec, GitReference, GitSpec, PathSpec, PixiSpec, UrlSpec};

#[serde_as]
#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
#[serde(rename_all = "kebab-case")]
struct RawSpec {
    /// The version spec of the package (e.g. `1.2.3`, `>=1.2.3`, `1.2.*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub version: Option<VersionSpec>,

    /// The URL of the package
    pub url: Option<Url>,

    /// The git url of the package
    pub git: Option<Url>,

    /// The path to the package
    pub path: Option<String>,

    /// The git revision of the package
    pub branch: Option<String>,

    /// The git revision of the package
    pub rev: Option<String>,

    /// The git revision of the package
    pub tag: Option<String>,

    /// The build string of the package (e.g. `py37_0`, `py37h6de7cb9_0`, `py*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub build: Option<StringMatcher>,

    /// The build number of the package
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub build_number: Option<BuildNumberSpec>,

    /// Match the specific filename of the package
    pub file_name: Option<String>,

    /// The channel of the package
    pub channel: Option<NamedChannelOrUrl>,

    /// The subdir of the channel
    pub subdir: Option<String>,

    /// The md5 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Md5>>")]
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Sha256>>")]
    pub sha256: Option<Sha256Hash>,
}

/// Returns a more helpful message when a version spec is used incorrectly.
fn version_spec_error<T: Into<String>>(input: T) -> Option<impl Display> {
    let input = input.into();
    if input.starts_with('/')
        || input.starts_with('.')
        || input.starts_with('\\')
        || input.starts_with("~/")
    {
        return Some(format!("it seems you're trying to add a path dependency, please specify as a table with a `path` key: '{{ path = \"{input}\" }}'"));
    }

    if input.contains("git") {
        return Some(format!("it seems you're trying to add a git dependency, please specify as a table with a `git` key: '{{ git = \"{input}\" }}'"));
    }

    if input.contains("://") {
        return Some(format!("it seems you're trying to add a url dependency, please specify as a table with a `url` key: '{{ url = \"{input}\" }}'"));
    }

    if let Ok(match_spec) = NamelessMatchSpec::from_str(&input, Lenient) {
        let spec = PixiSpec::from_nameless_matchspec(
            match_spec,
            &ChannelConfig::default_with_root_dir(PathBuf::default()),
        );
        return Some(format!(
            "expected a version specifier but looks like a matchspec, did you mean {}?",
            spec.to_toml_value()
        ));
    };

    if input.contains("subdir") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `subdir` key: '{ version = \"<VERSION_SPEC>\", subdir = \"<SUBDIR>\" }'".to_string());
    }

    if input.contains("channel") || input.contains("::") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `channel` key: '{ version = \"<VERSION_SPEC>\", channel = \"<CHANNEL>\" }'".to_string());
    }

    if input.contains("md5") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `md5` key: '{ version = \"<VERSION_SPEC>\", md5 = \"<MD5>\" }'".to_string());
    }

    if input.contains("sha256") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `sha256` key: '{ version = \"<VERSION_SPEC>\", sha256 = \"<SHA256>\" }'".to_string());
    }

    None
}

impl<'de> Deserialize<'de> for PixiSpec {
    fn deserialize<D>(deserializer: D) -> Result<PixiSpec, D::Error>
    where
        D: Deserializer<'de>,
    {
        serde_untagged::UntaggedEnumVisitor::new()
            .expecting(
                "a version string like \">=0.9.8\" or a detailed dependency like { version = \">=0.9.8\" }",
            )
            .string(|str| {
                VersionSpec::from_str(str, Strict)
                    .map_err(|_err|{
                        if let Some(msg) = version_spec_error(str) {
                            serde_untagged::de::Error::custom(msg)
                        } else {
                            serde_untagged::de::Error::custom("invalid version specifier")
                        }
                                            })
                    .map(PixiSpec::Version)
            })
            .map(|map| {
                let raw_spec: RawSpec = map.deserialize()?;

                if raw_spec.git.is_none()
                    && (raw_spec.branch.is_some()
                        || raw_spec.rev.is_some()
                        || raw_spec.tag.is_some())
                {
                    return Err(serde_untagged::de::Error::custom(
                        "`branch`, `rev`, and `tag` are only valid when `git` is specified",
                    ));
                }

                let is_git = raw_spec.git.is_some();
                let is_path = raw_spec.path.is_some();
                let is_url = raw_spec.url.is_some();


                let git_key = is_git.then_some("`git`");
                let path_key = is_path.then_some("`path`");
                let url_key = is_url.then_some("`url`");
                let non_detailed_keys = [git_key, path_key, url_key]
                    .into_iter()
                    .flatten()
                    .collect::<Vec<_>>()
                    .join(", ");

                if !non_detailed_keys.is_empty() && raw_spec.version.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`version` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.build.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.build_number.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.file_name.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.channel.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.subdir.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                let non_url_keys = [git_key, path_key]
                    .into_iter()
                    .flatten()
                    .collect::<Vec<_>>()
                    .join(", ");
                if !non_url_keys.is_empty() && raw_spec.sha256.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`sha256` cannot be used with {non_url_keys}"),
                    ));
                }
                if !non_url_keys.is_empty() && raw_spec.md5.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`md5` cannot be used with {non_url_keys}"),
                    ));
                }

                let spec = match (raw_spec.url, raw_spec.path, raw_spec.git) {
                    (Some(url), None, None) => PixiSpec::Url(UrlSpec {
                        url,
                        md5: raw_spec.md5,
                        sha256: raw_spec.sha256,
                    }),
                    (None, Some(path), None) => PixiSpec::Path(PathSpec { path: path.into() }),
                    (None, None, Some(git)) => {
                        let rev = match (raw_spec.branch, raw_spec.rev, raw_spec.tag) {
                            (Some(branch), None, None) => Some(GitReference::Branch(branch)),
                            (None, Some(rev), None) => Some(GitReference::Rev(rev)),
                            (None, None, Some(tag)) => Some(GitReference::Tag(tag)),
                            (None, None, None) => None,
                            _ => {
                                return Err(serde_untagged::de::Error::custom(
                                    "only one of `branch`, `rev`, or `tag` can be specified",
                                ));
                            }
                        };
                        PixiSpec::Git(GitSpec { git, rev })
                    },
                    (None, None, None) => {
                        let is_detailed =
                            raw_spec.version.is_some() ||
                                raw_spec.build.is_some() ||
                                raw_spec.build_number.is_some() ||
                                raw_spec.file_name.is_some() ||
                                raw_spec.channel.is_some() ||
                                raw_spec.subdir.is_some() ||
                                raw_spec.md5.is_some() ||
                                raw_spec.sha256.is_some();
                        if !is_detailed {
                            return Err(serde_untagged::de::Error::custom(
                                "one of `version`, `build`, `build-number`, `file-name`, `channel`, `subdir`, `md5`, `sha256`, `git`, `url`, or `path` must be specified",
                            ))
                        }

                        PixiSpec::DetailedVersion(DetailedSpec {
                            version: raw_spec.version,
                            build: raw_spec.build,
                            build_number: raw_spec.build_number,
                            file_name: raw_spec.file_name,
                            channel: raw_spec.channel,
                            subdir: raw_spec.subdir,
                            md5: raw_spec.md5,
                            sha256: raw_spec.sha256,
                        })
                    }
                    (_, _, _) => {
                        return Err(serde_untagged::de::Error::custom(
                            "only one of `url`, `path`, or `git` can be specified",
                        ))
                    }
                };

                Ok(spec)
            })
            .deserialize(deserializer)
    }
}

impl<'de> Deserialize<'de> for PathSpec {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct Raw {
            path: String,
        }

        Raw::deserialize(deserializer).map(|raw| PathSpec {
            path: raw.path.into(),
        })
    }
}

impl Serialize for PathSpec {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Raw {
            path: String,
        }

        Raw {
            path: self.path.to_string(),
        }
        .serialize(serializer)
    }
}

#[cfg(test)]
mod test {
    use serde::Serialize;
    use serde_json::{json, Value};

    use super::*;

    #[test]
    fn test_round_trip() {
        let examples = [
            json! { "1.2.3" },
            json!({ "version": "1.2.3" }),
            json!({ "version": "1.2.3", "build-number": ">=3" }),
            json! { "*" },
            json!({ "path": "foobar" }),
            json!({ "path": "~/.cache" }),
            json!({ "subdir": "linux-64" }),
            json!({ "channel": "conda-forge", "subdir": "linux-64" }),
            json!({ "channel": "conda-forge", "subdir": "linux-64" }),
            json!({ "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "version": "1.2.3", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main" }),
            // Errors:
            json!({ "ver": "1.2.3" }),
            json!({ "path": "foobar", "version": "1.2.3" }),
            json!({ "version": "//" }),
            json!({ "path": "foobar", "version": "//" }),
            json!({ "path": "foobar", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main", "tag": "v1" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json! { "/path/style"},
            json! { "./path/style"},
            json! { "\\path\\style"},
            json! { "~/path/style"},
            json! { "https://example.com"},
            json! { "https://github.com/conda-forge/21cmfast-feedstock"},
            json! { "1.2.3[subdir=linux-64]"},
            json! { "1.2.3[channel=conda-forge]"},
            json! { "conda-forge::1.2.3"},
            json! { "1.2.3[md5=315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3]"},
            json! { "1.2.3[sha256=315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3]"},
            json! { "*cpu*"},
            json! { "*=*openblas"},
        ];

        #[derive(Serialize)]
        struct Snapshot {
            input: Value,
            result: Value,
        }

        let mut snapshot = Vec::new();
        for input in examples {
            let spec: Result<PixiSpec, _> = serde_json::from_value(input.clone());
            let result = match spec {
                Ok(spec) => serde_json::to_value(&spec).unwrap(),
                Err(e) => {
                    json!({
                        "error": format!("ERROR: {e}")
                    })
                }
            };

            snapshot.push(Snapshot { input, result });
        }

        insta::assert_yaml_snapshot!(snapshot);
    }
}
