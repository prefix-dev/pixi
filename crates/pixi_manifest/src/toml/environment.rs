use crate::utils::PixiSpanned;
use serde::{Deserialize, Deserializer};

/// Helper struct to deserialize the environment from TOML.
/// The environment description can only hold these values.
#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub struct TomlEnvironment {
    #[serde(default)]
    pub features: PixiSpanned<Vec<String>>,
    pub solve_group: Option<String>,
    #[serde(default)]
    pub no_default_feature: bool,
}

#[derive(Debug)]
pub enum TomlEnvironmentList {
    Map(TomlEnvironment),
    Seq(Vec<String>),
}

impl<'de> Deserialize<'de> for TomlEnvironmentList {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        serde_untagged::UntaggedEnumVisitor::new()
            .map(|map| map.deserialize().map(TomlEnvironmentList::Map))
            .seq(|seq| seq.deserialize().map(TomlEnvironmentList::Seq))
            .expecting("either a map or a sequence")
            .deserialize(deserializer)
    }
}
