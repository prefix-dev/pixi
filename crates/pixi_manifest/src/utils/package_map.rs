use crate::utils::PixiSpanned;
use indexmap::IndexMap;
use pixi_spec::PixiSpec;
use serde::{
    de::{DeserializeSeed, MapAccess, Visitor},
    Deserialize, Deserializer, Serialize,
};
use std::ops::Range;
use std::{fmt, marker::PhantomData};

#[derive(Clone, Default, Debug, Serialize)]
pub struct UniquePackageMap {
    #[serde(flatten)]
    pub specs: IndexMap<rattler_conda_types::PackageName, PixiSpec>,

    #[serde(skip)]
    pub name_spans: IndexMap<rattler_conda_types::PackageName, Range<usize>>,

    #[serde(skip)]
    pub value_spans: IndexMap<rattler_conda_types::PackageName, Range<usize>>,
}

impl From<UniquePackageMap> for IndexMap<rattler_conda_types::PackageName, PixiSpec> {
    fn from(value: UniquePackageMap) -> Self {
        value.specs
    }
}

impl IntoIterator for UniquePackageMap {
    type Item = (rattler_conda_types::PackageName, PixiSpec);
    type IntoIter = indexmap::map::IntoIter<rattler_conda_types::PackageName, PixiSpec>;

    fn into_iter(self) -> Self::IntoIter {
        self.specs.into_iter()
    }
}

impl<'de> Deserialize<'de> for UniquePackageMap {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct PackageMapVisitor(PhantomData<()>);

        impl<'de> Visitor<'de> for PackageMapVisitor {
            type Value = UniquePackageMap;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                write!(formatter, "a map")
            }

            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
            where
                A: MapAccess<'de>,
            {
                let mut result = UniquePackageMap::default();
                while let Some((package_name, spec)) = map.next_entry_seed::<PackageMap, _>(
                    PackageMap(&result.specs),
                    PhantomData::<PixiSpanned<PixiSpec>>,
                )? {
                    let PixiSpanned {
                        span: package_name_span,
                        value: package_name,
                    } = package_name;
                    let PixiSpanned {
                        span: spec_span,
                        value: spec,
                    } = spec;
                    if let Some(package_name_span) = package_name_span {
                        result
                            .name_spans
                            .insert(package_name.clone(), package_name_span);
                    }
                    if let Some(spec_span) = spec_span {
                        result.value_spans.insert(package_name.clone(), spec_span);
                    }
                    result.specs.insert(package_name, spec);
                }

                Ok(result)
            }
        }
        let visitor = PackageMapVisitor(PhantomData);
        deserializer.deserialize_map(visitor)
    }
}

struct PackageMap<'a>(&'a IndexMap<rattler_conda_types::PackageName, PixiSpec>);

impl<'de, 'a> DeserializeSeed<'de> for PackageMap<'a> {
    type Value = PixiSpanned<rattler_conda_types::PackageName>;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        let package_name = Self::Value::deserialize(deserializer)?;
        match self.0.get_key_value(&package_name.value) {
            Some((package_name, _)) => {
                Err(serde::de::Error::custom(
                    format!(
                        "duplicate dependency: {} (please avoid using capitalized names for the dependencies)", package_name.as_source())
                ))
            }
            None => Ok(package_name),
        }
    }
}
