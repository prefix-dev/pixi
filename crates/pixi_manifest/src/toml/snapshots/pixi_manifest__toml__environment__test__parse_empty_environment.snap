---
source: crates/pixi_manifest/src/toml/environment.rs
expression: "format_parse_error(input, TopLevel::from_toml_str(input).unwrap_err())"
---
  × missing field 'features' in table
   ╭─[pixi.toml:2:19]
 1 │
 2 │             env = {}
   ·                   ──
 3 │
   ╰────
